### PR TITLE
Rework event APIs to provide threadsafe defer method

### PR DESCRIPTION
**Pull request by grgustaf**

### CODE REVIEW

**Comment by grgustaf** [comment]
Here are the areas that changed w/ this patch and a checklist for testing them; probably don't need 100% coverage to merge but just FYI.

- [ ] ble
- [x] net
  -  [x] eth w/ ipv4
  -  [x] eth w/ ipv6
  -  [x] ble w/ 6lowpan
- [ ] ocf client
- [ ] ocf server
- [x] uart
- [ ] websockets
  -  [ ] eth w/ ipv4
  -  [ ] eth w/ ipv6
  -  [x] ble w/ 6lowpan

**Comment by grgustaf** [comment]
@haoxli not sure where you are in the test cycle, but if @jimmy-huang gets this one merged soon it would be great if we could test w/ this ASAP since it's a big change.

File: src/zjs_net.c [hunk 1]
```diff
@@ -123,13 +120,19 @@ typedef struct sock_handle {
...
-static const jerry_object_native_info_t net_type_info = {
-   .free_cb = free_handle_nop
-};
+// get the socket handle or return a JS error
+#define GET_SOCK_HANDLE_JS(obj, var)                                       \
+    sock_handle_t *var = (sock_handle_t *)zjs_event_get_user_handle(obj);  \
+    if (!var) { return zjs_error("no socket handle"); }
+
+// get the net handle or return a JS error
+#define GET_NET_HANDLE_JS(obj, var)                                       \
```
Comment (jimmy-huang) [inline src/zjs_net.c]: alight \

File: src/zjs_ble.c [hunk 2]
```diff
@@ -827,11 +794,17 @@ static ZJS_DECL_FUNC(zjs_ble_start_advertising)
...
-
-    zjs_trigger_event(ble_handle->ble_obj, "advertisingStart", &error, 1, NULL,
-                      NULL);
-    DBG_PRINT("BLE event: adveristingStart\n");
+    jerry_value_t error;
+    if (err) {
+        error = zjs_error("advertising failed");
+        jerry_value_clear_error_flag(&error);
+    }
+    else {
```
Comment (jimmy-huang) [inline src/zjs_ble.c]: } else {

File: src/zjs_net.c [hunk 3]
```diff
@@ -255,14 +228,99 @@ static void post_closed(void *handle)
...
+                            const char *buffer, u32_t bytes)
+{
+    sock_handle_t *handle = (sock_handle_t *)h;
+
+    // find length of unconsumed data in read buffer
+    u32_t len = handle->wptr - handle->rptr;
+    zjs_buffer_t *zbuf;
+    jerry_value_t data_buf = zjs_buffer_create(len, &zbuf);
+    if (!zbuf) {
+        // out of memory
```
Comment (jimmy-huang) [inline src/zjs_net.c]: print error?

File: src/zjs_ws.json [hunk 4]
```diff
@@ -15,7 +15,7 @@
             "CONFIG_NET_STATISTICS=y",
             "CONFIG_NET_BUF=y",
             "CONFIG_NET_PKT_RX_COUNT=5",
-            "CONFIG_NET_PKT_TX_COUNT=2",
+            "CONFIG_NET_PKT_TX_COUNT=5",
             "CONFIG_NET_BUF_RX_COUNT=10",
             "CONFIG_NET_BUF_TX_COUNT=6",
```
Comment (jimmy-huang) [inline src/zjs_ws.json]: you might change the TX_COUNT to 10 as well to match the RX_COUNT?

Comment (grgustaf) [inline src/zjs_ws.json]: It matches the pkt_rx_count...

Comment (grgustaf) [inline src/zjs_ws.json]: Really it makes sense for TX and RX to not match though (as in buf count) because we have control over TX but not so much RX.

**Comment by jimmy-huang** [comment]
+1

**Comment by haoxli** [comment]
@grgustaf We've completed the testing of this week. But for this change, we will run an additional testing for the impacted modules in the next two days.

**Comment by qiaojingx** [comment]
Finished testing with commit 4d9b084 for BLE, net, UART, ocf and websocket modules.
The pass rate of above modules have upgraded on A101, K64F and Linux with this PR #1420, but still found two new issues:  #1421, #1422.
Fixed issues: #1386, #1369, #1407, #1234, #1261.

**Pull request merged**